### PR TITLE
ER-888 Implements select employer for Provider

### PR DESCRIPTION
**Pull request by dev4valtech**
I have created a partial class for VacancyClient to implement IProviderVacancyClient.

Ignore bit around Title.... this will change when I come to that story. 

Also ignore hard coded data return of ProviderEditVacancyInfo, this will change when I will do the job task.

### CODE REVIEW

File: src/Provider/Provider.Web/ViewModels/Part1/Employer/EmployersViewModel.cs [hunk 1]
```diff
@@ -0,0 +1,11 @@
+using System.Collections.Generic;
+using Esfa.Recruit.Shared.Web.ViewModels;
+
+namespace Esfa.Recruit.Provider.Web.ViewModels.Part1.Employer
+{
+    public class EmployersViewModel : EmployersEditViewModel
```
Comment (chrisjensenuk) [inline src/Provider/Provider.Web/ViewModels/Part1/Employer/EmployersViewModel.cs]: Apparently we're not inheriting from the edit model any more so remove this dependency and copy over the required params.

File: src/Provider/Provider.Web/RouteModel/VacancyRouteModel.cs [hunk 2]
```diff
@@ -5,16 +5,10 @@ namespace Esfa.Recruit.Provider.Web.RouteModel
...
     {
-        private string _ukprn;
-
         [FromRoute]
-        public string Ukprn
-        {
-            get => _ukprn;
-            set => _ukprn = value.ToUpper();
-        }
+        public long Ukprn { get; set; }
```
Comment (chrisjensenuk) [inline src/Provider/Provider.Web/RouteModel/VacancyRouteModel.cs]: Should this be string so we don't get model binder type issues?

File: src/Shared/Recruit.Vacancies.Client/Infrastructure/Client/VacancyClient.cs [hunk 3]
```diff
@@ -21,7 +20,7 @@
 
 namespace Esfa.Recruit.Vacancies.Client.Infrastructure.Client
 {
-    public class VacancyClient : IRecruitVacancyClient, IEmployerVacancyClient, IProviderVacancyClient, IJobsVacancyClient
+    public partial class VacancyClient : IRecruitVacancyClient, IEmployerVacancyClient, IJobsVacancyClient
```
Comment (chrisjensenuk) [inline src/Shared/Recruit.Vacancies.Client/Infrastructure/Client/VacancyClient.cs]: partial?

File: src/Provider/Provider.Web/Orchestrators/Part1/EmployerOrchestrator.cs [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+
+                return await ValidateAndExecute(
+                    newVacancy, 
+                    v => _recruitVacancyClient.Validate(v, ValidationRules),
+                    async v => await _providerVacancyClient.CreateVacancyAsync(SourceOrigin.ProviderWeb, 
+                        viewModel.SelectedEmployerId, viewModel.Ukprn, user, UserType.Provider));
+            }
+
+            var vacancy = await Utility.GetAuthorisedVacancyForEditAsync(_providerVacancyClient, _recruitVacancyClient, 
+                new VacancyRouteModel{ Ukprn = viewModel.Ukprn, VacancyId = viewModel.VacancyId.GetValueOrDefault() },
```
Comment (chrisjensenuk) [inline src/Provider/Provider.Web/Orchestrators/Part1/EmployerOrchestrator.cs]: Recently I've been adding the VRM as an extra parameter to the controller action and passing it down to the orchestrator so I don't have to recreate it.

For example: https://github.com/SkillsFundingAgency/das-recruit/blob/c9d054253eb5566f9d31fcd59454ef10aa15a853/src/Employer/Employer.Web/Orchestrators/Part2/QualificationsOrchestrator.cs#L71

https://github.com/SkillsFundingAgency/das-recruit/blob/c9d054253eb5566f9d31fcd59454ef10aa15a853/src/Employer/Employer.Web/Controllers/Part2/QualificationsController.cs#L36

File: src/Provider/Provider.Web/Orchestrators/Part1/EmployerOrchestrator.cs [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+            var vm = new EmployersViewModel
+            {
+                PageInfo = new PartOnePageInfoViewModel(),
+                Employers = editVacancyInfo.Employers.Select(e => new EmployerViewModel {Id = e.Id, Name = e.Name})
+            };
+
+            return vm;
+        }
+
+        public async Task<OrchestratorResponse<Guid>> PostEmployerEditViewModelAsync(EmployersEditViewModel viewModel, VacancyUser user)
```
Comment (chrisjensenuk) [inline src/Provider/Provider.Web/Orchestrators/Part1/EmployerOrchestrator.cs]: `PostEmployerEditModelAsync`

**Comment by dev4valtech** [comment]
@chrisjensenuk  are you suggesting that we do not create the vacancy on step 1 (Employer select) and create after step 2 (Title)? In which case it makes sense to have the title at the least.

File: src/Provider/Provider.Web/RouteModel/VacancyRouteModel.cs (hunk 2 above)
Comment (dev4valtech) [reply src/Provider/Provider.Web/RouteModel/VacancyRouteModel.cs]: Thinking about this, we should be reading it from the user, but I am not sure if that is possible.

**Comment by chrisjensenuk** [comment]
@dev4valtech I think that's what I'm getting at, yes.  Maybe one for a follow up ticket. Can you liaise with Mark about this issue?

**Comment by mgwilliam** [comment]
@chrisjensenuk @dev4valtech I think we should only create the vacancy once we have the title (so it's basically the same as when we create for an employer)

File: src/Shared/Recruit.Vacancies.Client/Infrastructure/Client/VacancyClient.cs (hunk 3 above)
Comment (dev4valtech) [reply src/Shared/Recruit.Vacancies.Client/Infrastructure/Client/VacancyClient.cs]: I think this one has potential to become as big, so just watch the space grow. ;-)

File: src/Provider/Provider.Web/Orchestrators/Part1/EmployerOrchestrator.cs (hunk 5 above)
Comment (dev4valtech) [reply src/Provider/Provider.Web/Orchestrators/Part1/EmployerOrchestrator.cs]: Not very happy with these naming convention. I think a more appropriate would be PostEmployerSelectionAsync. Fixing it anyway for "consistency".

File: src/Jobs/Recruit.Vacancies.Jobs/DomainEvents/Handlers/Provider/SetupProviderHandler.cs [hunk 6]
```diff
@@ -8,13 +8,15 @@
 
 namespace Esfa.Recruit.Vacancies.Jobs.DomainEvents.Handlers.Provider
 {
-    public class SetupProviderDomainEventHandler : DomainEventHandler,  IDomainEventHandler<SetupProviderEvent>
+    public class SetupProviderHandler : DomainEventHandler,  IDomainEventHandler<SetupProviderEvent>
     {
-        private readonly ILogger<SetupProviderDomainEventHandler> _logger;
```
[the remark this answers is not in the document]
Comment (dev4valtech) [reply src/Jobs/Recruit.Vacancies.Jobs/DomainEvents/Handlers/Provider/SetupProviderHandler.cs]: I have just made it "consistent" with other handlers. But I agree with your idea. May be we need to agree on this as a team and apply it everywhere so it stays "consistent".

File: src/Provider/Provider.Web/Orchestrators/Part1/TitleOrchestrator.cs [hunk 7]
```diff
@@ -59,49 +59,14 @@ public async Task<TitleViewModel> GetTitleViewModelAsync(VacancyRouteModel vrm)
...
-                    v => _vacancyClient.Validate(v, ValidationRules),
-                    async v => await _client.CreateVacancyAsync(SourceOrigin.ProviderWeb, m.Title, numberOfPositions.Value, long.Parse(m.Ukprn), user, UserType.Provider));
-            }
-
             var vacancy = await Utility.GetAuthorisedVacancyForEditAsync(_client, _vacancyClient, 
                 new VacancyRouteModel{Ukprn = m.Ukprn, VacancyId = m.VacancyId.Value}, RouteNames.Title_Post);
 
             vacancy.Title = m.Title;
+
+            var numberOfPositions = int.TryParse(m.NumberOfPositions, out var n)? n : default(int?);
```
[the remark this answers is not in the document]
Comment (dev4valtech) [reply src/Provider/Provider.Web/Orchestrators/Part1/TitleOrchestrator.cs]: Ignore, not sure where this came from. I will fix this as part of Title related task.

File: src/Provider/Provider.Web/ViewModels/Part1/Employer/EmployersEditViewModel.cs [hunk 8]
```diff
@@ -0,0 +1,11 @@
+using System;
+using Esfa.Recruit.Provider.Web.RouteModel;
+using Microsoft.AspNetCore.Mvc;
+
+namespace Esfa.Recruit.Provider.Web.ViewModels.Part1.Employer
+{
+    public class EmployersEditViewModel : VacancyRouteModel
```
[the remark this answers is not in the document]
Comment (dev4valtech) [reply src/Provider/Provider.Web/ViewModels/Part1/Employer/EmployersEditViewModel.cs]: what do you mean?

File: src/Provider/Provider.Web/ViewModels/Part1/Title/TitleEditModel.cs [hunk 9]
```diff
@@ -8,7 +8,7 @@ namespace Esfa.Recruit.Provider.Web.ViewModels.Part1.Title
     public class TitleEditModel
     {
         [FromRoute]
-        public string Ukprn { get; set; }
+        public long Ukprn { get; set; }
 
         [FromRoute]
         public Guid? VacancyId { get; set; }
```
[the remark this answers is not in the document]
Comment (dev4valtech) [reply src/Provider/Provider.Web/ViewModels/Part1/Title/TitleEditModel.cs]: I will fix it as part of Title story

File: src/Provider/Provider.Web/Orchestrators/Part1/EmployerOrchestrator.cs [hunk 10]
```diff
@@ -0,0 +1,96 @@
...
+using Microsoft.Extensions.Logging;
+using Esfa.Recruit.Vacancies.Client.Infrastructure.Client;
+using Esfa.Recruit.Provider.Web.Configuration.Routing;
+using Esfa.Recruit.Vacancies.Client.Infrastructure.QueryStore.Projections.EditVacancyInfo;
+
+namespace Esfa.Recruit.Provider.Web.Orchestrators.Part1
+{
+    public class EmployerOrchestrator : EntityValidatingOrchestrator<Vacancy, EmployersEditViewModel>
+    {
+        private const VacancyRuleSet ValidationRules = VacancyRuleSet.EmployerAccountId | VacancyRuleSet.TrainingProvider;
```
[the remark this answers is not in the document]
Comment (dev4valtech) [reply src/Provider/Provider.Web/Orchestrators/Part1/EmployerOrchestrator.cs]: You are right, I will remove it from here. It is being validate else where so we are good.

File: src/Provider/Provider.Web/Views/Shared/_Layout.cshtml [hunk 11]
```diff
@@ -130,9 +130,6 @@
             </p>
         </div>
         <main class="govuk-main-wrapper " id="main-content" role="main">
-
-            Your training provider account @User.GetUkprn() @User.GetDisplayName()
```
[the remark this answers is not in the document]
Comment (dev4valtech) [reply src/Provider/Provider.Web/Views/Shared/_Layout.cshtml]: do we still need this?

File: src/Shared/Recruit.Vacancies.Client/Infrastructure/Client/IProviderVacancyClient.cs [hunk 12]
```diff
@@ -1,17 +1,18 @@
...
 using Esfa.Recruit.Vacancies.Client.Domain.Entities;
+using Esfa.Recruit.Vacancies.Client.Infrastructure.QueryStore.Projections.EditVacancyInfo;
 using Esfa.Recruit.Vacancies.Client.Infrastructure.QueryStore.Projections.Provider;
 
 namespace Esfa.Recruit.Vacancies.Client.Infrastructure.Client
 {
     public interface IProviderVacancyClient
     {
-        Task<Guid> CreateVacancyAsync(SourceOrigin origin, string title, int numberOfPositions, long ukprn, VacancyUser user, UserType userType);
+        Task<Guid> CreateVacancyAsync(SourceOrigin origin, string employerAccountId, long ukprn, VacancyUser user, UserType userType);
```
[the remark this answers is not in the document]
Comment (dev4valtech) [reply src/Shared/Recruit.Vacancies.Client/Infrastructure/Client/IProviderVacancyClient.cs]: Also API would be a complete Vacancy object... so not worried about it here.

**Pull request merged**